### ISSUE TITLE
Graphics state is broken after resume from suspend-to-RAM

### ISSUE THREAD

**Issue by anderspapitto** [NixOS/nixpkgs#6692]
Suspend-to-RAM is unusably broken for me with NixOS. When booting into Ubuntu 14.04 from a live USB on the same hardware, suspend-to-RAM works perfectly.

CONFIGURATION
I am running on a 3rd generation Lenovo X1 Carbon. My configuration.nix is here https://gist.github.com/anderspapitto/1454c0dc54a3eea69b60, and my hardware-configuration.nix is here https://gist.github.com/anderspapitto/04bf411c2c919bc89617. I am on nixpkgs-unstable, at a commit described by 

```
[root@gurney:~]# cat /root/.nix-defexpr/channels/manifest.nix 
[ { meta = { }; name = "nixos-15.05pre58357.a869c83"; out = { outPath = "/nix/store/0fq4hfpq3pxcakal4bcpn1znbyawp1w3-nixos-15.05pre58357.a869c83"; }; outPath = "/nix/store/0fq4hfpq3pxcakal4bcpn1znbyawp1w3-nixos-15.05pre58357.a869c83"; outputs = [ "out" ]; system = "x86_64-linux"; type = "derivation"; } ]
```

PROBLEM DESCRIPTION
When I suspend (using `systemctl suspend`, either as my unprivileged user or as root), and then resume by pressing the power button, it appears that something is broken in the graphics state. My display manager (slim) shows up correctly, except that when I enter my password, I don't see dots appear in the password box, and there is no cursor. After logging in, my status bar is missing, there is some pixelated garbage in the top left corner of the screen, and many operations don't work. I can switch workspaces, but can't launch programs or interact with running windows.

Interestingly, restarting the X server (with `sudo systemctl restart display-manager`) does NOT fix this issue. At the same time, all vttys other than the one with X running in it are unaffected. Even more interestingly, if I turn off X (`sudo systemctl stop display-manager`), suspend and resume, and then start X afterwards, it displays the same symptoms. Only a reboot helps.

LOGS
I have tried looking in journalctl output, but I don't see anything useful. I see some errors and warnings, but these are the same when first booting as when resuming from suspend or when restarting display-manager. Here is the output of `journalctl -b -1`
https://gist.github.com/anderspapitto/3841395e3c95b167dcab

I'm happy to gather any more information and try out suggestions, but I'm not sure how to continue investigating this on my own.

**Comment by codyopel** [comment]
Duplicate of #6617

**Comment by cillianderoiste** [comment]
@codyopel #6617 isn't related (It's about configuring the resumeDevice for hibernation).

@anderspapitto I don't see any mention of services.xserver.videoDrivers, did you try setting that correctly? I guess it has an intel GPU? Otherwise, have you found reports of this issue on other distros?

**Comment by vcunat** [comment]
I would expect that something around the graphic driver stack is causing it. I see a potentially problematic message during suspend:
```
[drm:stop_ring [i915]] *ERROR* render ring : timed out trying to stop ring
```

``/var/log/X.0.log`` might show more interesting things, and I also suspect that setting ``services.xserver.videoDrivers = [ "intel" ];`` might get you a better X driver and solve this.

**Comment by anderspapitto** [comment]
I have an integrated intel gpu (Intel HD Graphics 5500). Setting `services.xserver.videoDrivers` does not fix the issue - it looks like `"intel"` is in the default list, so it was probably being used already.

I don't see anything relevant in the X logs. Here is an example
https://gist.github.com/anderspapitto/a5d1e6207226634b3137

@vcunat it seems like that error has been seen related to similar issues. I'll try following this discussion
http://markmail.org/thread/nnibszsh3pcca7gg#query:+page:1+mid:jaiyfhwxzhp52loa+state:results

**Comment by anderspapitto** [comment]
It looks the workaround from that thread (run glxgears before suspending) works for me, and the fix is already upstream in the 4.0 kernel, so I'll close this for now. Thanks!
